rust/lance-encoding/src/decoder.rs

```diff
@@ -1285,6 +1304,137 @@ impl BatchDecodeStream {
+    requested_rows: RequestedRows,
+    filter: FilterExpression,
+    column_indices: Vec<u32>,
+    target_schema: Arc<Schema>,
+    config: SchedulerDecoderConfig,
+) -> Result<BoxStream<'static, ReadBatchTask>> {
+    let num_rows = requested_rows.num_rows();
+
+    let mut decode_scheduler = DecodeBatchScheduler::try_new(
+        target_schema.as_ref(),
+        &column_indices,
+        &column_infos,
+        &vec![],
+        num_rows,
+        config.decoder_strategy,
+        config.io.clone(),
+        config.cache,
+        &filter,
+    )
+    .await?;
+
+    let root_decoder = match &requested_rows {
+        RequestedRows::Ranges(ranges) => decode_scheduler.new_root_decoder_ranges(ranges),
+        RequestedRows::Indices(indices) => decode_scheduler.new_root_decoder_indices(indices),
+    };
+
+    let (tx, rx) = mpsc::unbounded_channel();
+
+    let io = config.io;
+    let scheduler_handle = tokio::task::spawn(async move {
+        match requested_rows {
+            RequestedRows::Ranges(ranges) => {
+                decode_scheduler.schedule_ranges(&ranges, &filter, tx, io)
+            }
+            RequestedRows::Indices(indices) => {
+                decode_scheduler.schedule_take(&indices, &filter, tx, io)
+            }
+        }
+    });
+
+    let decode_stream =
+        BatchDecodeStream::new(rx, config.batch_size, num_rows, root_decoder).into_stream();
+
+    Ok(check_scheduler_on_drop(decode_stream, scheduler_handle))
+}
+
+/// Launches a scheduler on a dedicated (spawned) task and creates a decoder to
+/// decode the scheduled data and returns the decoder as a stream of record batches.
+///
+/// This is a convenience function that creates both the scheduler and the decoder
+/// which can be a little tricky to get right.
+pub fn schedule_and_decode(
+    column_infos: Vec<Arc<ColumnInfo>>,
+    requested_rows: RequestedRows,
+    filter: FilterExpression,
+    column_indices: Vec<u32>,
+    target_schema: Arc<Schema>,
+    config: SchedulerDecoderConfig,
+) -> BoxStream<'static, ReadBatchTask> {
+    if requested_rows.num_rows() == 0 {
+        return stream::empty().boxed();
+    }
+    // For convenience we really want this method to be a snchronous method where all
+    // errors happen on the stream.  There is some async initialization that must happen
+    // when creating a scheduler.  We wrap that all up in the very first task.
+    stream::once(create_scheduler_decoder(
+        column_infos,
+        requested_rows,
+        filter,
+        column_indices,
+        target_schema,
+        config,
+    ))
+    .map(|maybe_stream| match maybe_stream {
+        // If the initialization failed make it look like a failed task
+        Ok(stream) => stream,
+        Err(e) => stream::once(std::future::ready(ReadBatchTask {
+            num_rows: 0,
+            task: std::future::ready(Err(e)).boxed(),
+        }))
+        .boxed(),
+    })
+    .flatten()
+    .boxed()
+}
+
```

Comment:
This code is _mostly_ a port of code that was in `FileReader` (there were actually two copies of it in `FileReader`).  It's been slightly changed to now call `initialize` after creating the `DecodeBatchScheduler`.